### DISCUSSION TITLE
Annotating subquery used in relationship with `foreign` or `remote`

### DISCUSSION THREAD

**Discussion by Gero-B** [sqlalchemy/sqlalchemy#10074] [Usage Questions] [answered]
Hi, I'm attempting to create the correct relationship join clause to solve the following scenario `youngest_item_sibling`:

```python
class Item(Base):
    __tablename__ = "items"

    id = Column(UUID, primary_key=True)
    parent_id = Column(UUID, ForeignKey("items.id"), nullable=True)
    external_id = Column(String, nullable=False)
    created_at = Column(DateTime, nullable=False)

class Related(Base):
    __tablename__ = "related"

    external_id = Column(String, nullable=False)

    related_item = relationship(Item, primaryjoin=Item.external_id==external_id, uselist=False)
    youngest_item_sibling = relationship(
        "Item",
        primaryjoin=lambda: Related.get_youngest_item_sibling_clause(),
        uselist=False,
        viewonly=True,
    )
    @classmethod
    def get_youngest_item_sibling_clause(cls):
        item1 = Item.__table__.alias("item1")
        item2 = Item.__table__.alias("item2")

        youngest_sibling_id = (
            select(item2.c.id)
            .join(
                item1,
                and_(
                    func.coalesce(item1.c.parent_id, item1.c.id) == func.coalesce(item2.c.parent_id, item2.c.id),
                    item1.c.external_id == cls.external_id,
                ),
            )
            .order_by(item2.c.created_at.desc())
            .limit(1)
            .scalar_subquery()
        )
        
        return Item.id == youngest_sibling_id

```

Notes:
- parent/child relationship is only one generation deep
- exactly one `related_item` always exists
- `youngest_item_sibling` shouldn't be None (hence the coalescing) we either return:
  - youngest sibling (younger than `related_item`)
  - youngest child (if `related_item` is parent with children)
  - current `related_item` if it is already the youngest sibling
  - current `related_item` if it is the parent with no children

**The problem:**
I'm unable to annotate the `primaryjoin` condition `youngest_sibling_id` with `foreign()` or `remote()` 

when I do `remote(youngest_sibling_id)` it produces the following error:
`TypeError: _annotate_columns.<locals>.clone() got an unexpected keyword argument 'replace'`

Is there a better way to define this relationship? or is this an issue with annotating subquery expressions?

**Comment by zzzeek** [comment, marked as answer]
your primaryjoin does not seem to refer to Related.id by itself on either side so I dont think there's any way it would work like that.   The primaryjoin should have two sides; one side is columns on Related, the other side are columns on Item.  There's not really a good way to try to force correlated subqueries in there, since this is ultimately a JOIN.

Since you are doing some very complicated SELECT for Item.id where it looks like you have to select from both item1.id (to match to Related) and item2.id (to match to the item you want) at the same time, the most flexible approach is relationship to aliased class documented at https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#relationship-to-aliased-class .    To work this into a pattern where you are selecting from an aggregate (in this case it will be the highest date; `func.max(item2.c.created_at)` the most efficient way is to follow the row-limited relationships pattern illustrated at https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#row-limited-relationships-with-window-functions .

**Reply by Gero-B** [reply]
Thanks! Will give this a shot
